desc/magnetic_fields.py

```diff
@@ -795,3 +831,464 @@ def odefun(rpz, s):
+                raise TypeError(
+                    "Expected type FourierRZToroidalSurface for argument surface, "
+                    f"instead got type {type(surface)}"
+                )
+            R_lmn = surface.R_lmn
+            Z_lmn = surface.Z_lmn
+            modes_R = surface._R_basis.modes[:, 1:]
+            modes_Z = surface._Z_basis.modes[:, 1:]
+            NFP = surface.NFP
+            sym = surface.sym
+            rho = surface.rho
+
+        assert surface_grid.NFP == NFP, "NFP of surface must match NFP of surface_grid"
+
+        super().__init__(
+            R_lmn, Z_lmn, modes_R, modes_Z, NFP, sym, rho, name, check_orientation
+        )
+        self._set_derivatives(potential_dtheta, potential_dzeta)
+
+    @property
+    def surface_grid(self):
+        """Grid of points to evaluate surface current at."""
+        return self._surface_grid
+
+    @surface_grid.setter
+    def surface_grid(self, new):
+        if new != self._surface_grid:
+            self._surface_grid = new
+
+    @property
+    def params(self):
+        """Dict of parameters to pass to potential function and its derivatives."""
+        return self._params
+
+    @params.setter
+    def params(self, new):
+        if new != self._params:
+            if len(new) != len(self._params):
+                raise UserWarning(
+                    "Length of new params is different from length of current params! "
+                    "May cause errors unless potential function is also changed."
+                )
+            self._params = new
+
+    @property
+    def potential(self):
+        """Potential function, signature (theta,zeta,**params) -> potential value."""
+        return self._potential
+
+    @potential.setter
+    def potential(self, new):
+        if new != self._potential:
+            self._potential = new
+            # reset derivatives if potential has changed, using AD
+            self._set_derivatives()
+
+    def _set_derivatives(self, potential_dtheta=None, potential_dzeta=None):
+        if potential_dtheta:
+            self._potential_t = partial(potential_dtheta, **self.params)
+        else:
+            self._potential_t = vmap(
+                Derivative(
+                    partial(self.potential, **self.params),
+                    argnum=0,
+                    mode="grad",
+                ).compute,
+                in_axes=[0, 0],
+            )
+        if potential_dzeta:
+            self._potential_z = partial(potential_dzeta, **self.params)
+        else:
+            self._potential_z = vmap(
+                Derivative(
+                    partial(self.potential, **self.params), argnum=1, mode="grad"
+                ).compute,
+                in_axes=[0, 0],
+            )
+
+    def _compute_surface_current(self, surface_grid=None, params=None):
+        if surface_grid is None:
+            surface_grid = self.surface_grid
+        if (params is None) or (len(params) == 0):
```

Comment:
might be better to leave out the `or len(params) == 0` bit? I think all the magnetic fields default to `params=None` at this point.

Reply:
also these params aren't passed to the potential functions